ethcore/private-tx/src/lib.rs

```diff
@@ -256,48 +259,109 @@ impl Provider where {
-						bail!(ErrorKind::ValidatorAccountNotSet);
+			let private_hash = transaction.private_transaction.hash();
+			match transaction.validator_account {
+				None => {
+					trace!("Propagating transaction further");
+					self.broadcast_private_transaction(private_hash, transaction.private_transaction.rlp_bytes().into_vec());
+					return Ok(())
+				}
+				Some(validator_account) => {
+					if !self.validator_accounts.contains(&validator_account) {
+						trace!("Propagating transaction further");
+						self.broadcast_private_transaction(private_hash, transaction.private_transaction.rlp_bytes().into_vec());
+						return Ok(())
 					}
-					let account = desc.validator_account;
-					if let Action::Call(contract) = transaction.signed().action {
+					let signed_tx = transaction.transaction.clone();
+					if let Action::Call(contract) = signed_tx.action {
 						// TODO [ToDr] Usage of BlockId::Latest
 						let contract_nonce = self.get_contract_nonce(&contract, BlockId::Latest)?;
-						let private_state = self.execute_private_transaction(BlockId::Latest, transaction.signed())?;
+						let private_state = self.execute_private_transaction(BlockId::Latest, &signed_tx)?;
 						let private_state_hash = self.calculate_state_hash(&private_state, contract_nonce);
 						trace!("Hashed effective private state for validator: {:?}", private_state_hash);
-						let password = find_account_password(&self.passwords, &*self.accounts, &account);
-						let signed_state = self.accounts.sign(account, password, private_state_hash)?;
-						let signed_private_transaction = SignedPrivateTransaction::new(desc.private_hash, signed_state, None);
+						let password = find_account_password(&self.passwords, &*self.accounts, &validator_account);
+						let signed_state = self.accounts.sign(validator_account, password, private_state_hash)?;
+						let signed_private_transaction = SignedPrivateTransaction::new(private_hash, signed_state, None);
 						trace!("Sending signature for private transaction: {:?}", signed_private_transaction);
-						self.broadcast_signed_private_transaction(signed_private_transaction.rlp_bytes().into_vec());
+						self.broadcast_signed_private_transaction(signed_private_transaction.hash(), signed_private_transaction.rlp_bytes().into_vec());
 					} else {
 						warn!("Incorrect type of action for the transaction");
 					}
-				},
-				Err(e) => {
-					warn!("Cannot retrieve descriptor for transaction with error {:?}", e);
 				}
 			}
-			verification_queue.remove_private_transaction(&transaction_hash);
 		}
 		Ok(())
 	}
 
+	/// Add signed private transaction into the store	/// Creates corresponding public transaction if last required signature collected and sends it to the chain
+	pub fn process_signatures(&self) -> Result<(), Error> {
+		trace!("Processing signed private transactions");
+		let mut signed_transactions = self.imported_signed_transactions.lock();
+		for tx in signed_transactions.drain(..) {
```

Comment:
Again error handling here:
We iterate over a list of transactions, but any error inside here (or `return` statement) will cause the draining iterator to be dropped and removing all elements anyway.

If the assumption is that there is always only one transaction then let's make it explicit and pass the transaction as function parameter as suggested in a comment above.